modules/migrations/gitea_uploader.go

```diff
@@ -690,6 +690,15 @@ func (g *GiteaLocalUploader) newPullRequest(pr *base.PullRequest) (*models.PullR
 		}
 	} else {
 		head = pr.Head.Ref
+		// Ensure the closed PR SHA still points to an existing ref
+		_, err = git.NewCommand("rev-list", "-1", pr.Head.SHA).RunInDir(g.repo.RepoPath())
+		if err != nil {
+			log.Warn("Deprecated local head, removing : %v", pr.Head.SHA)
+			err = os.Remove(filepath.Join(pullHead, "head"))
```

Comment:
Damn I think this may be the incorrect thing.

Reply:
You need to use `git update-ref -d`

Reply:
Hello and thank you for pointing this out ! (Damn, I learn everyday with git ! :D )
Indeed the update-ref [should do the trick, but I tried this workaround in my gist](https://gist.github.com/99rgosse/b4ac3782f1b3c837b0aa01690b5b7095)

```suggestion
			relPath := "refs/pull/" + fmt.Sprintf("%d", pr.Number) + "/head"
			_, err = git.NewCommand("update-ref", "-d", relPath).RunInDir(g.repo.RepoPath())
```

but I'm not happy with it  :( :
- I cannot specify system full path in update-ref -d (thus the ugly relPath variable I put)
- If the path does not exist, update-ref will not throw an error, just move on (tried locally)
- update-ref can keep invalid sha1 pointers :

> `git fsck --full --lost-found
Checking object directories: 100% (256/256), done.
Checking objects: 100% (165029/165029), done.
error: refs/pull/20/head: invalid sha1 pointer 0000000000000000000000000000000000000000
error: refs/pull/64/head: invalid sha1 pointer 0000000000000000000000000000000000000000`

Should I keep on and find another workaround for those invalid pointers ?
Thanks in advance...

Reply:
I tried it with a Git version 2.33 (instead of the Git 2.17 still on ubuntu 18) 
Now update-ref gives an error for those invalid refs, but still not delete them :
`git update-ref -d refs/pull/20/head
error: cannot lock ref 'refs/pull/20/head': unable to resolve reference 'refs/pull/20/head': reference broken`

Reply:
damn - the trouble is you can't just delete the file as the ref might be packed away. It may be that you could try using: 

```
git update-ref --no-deref -d refs/pull/20/head
```

Reply:
Hello and thank you ! 
 I tried --no-deref with no luck, we still keep these empty "head" files that are redirected to 40 '0'.
`git fsck --full --lost-found`
``
`Checking object directories: 100% (256/256), done.`
`Checking objects: 100% (165114/165114), done.`
``
`error: refs/pull/20/head: invalid sha1 pointer 0000000000000000000000000000000000000000`
`error: refs/pull/64/head: invalid sha1 pointer 0000000000000000000000000000000000000000`

instead, I modified to take in account these 0uid SHA1 references (that do not trigger update-ref - I checked the manpage)

Now, these 0uid were only from closed migrated PR with deleted branches or fork, so in anyway it won't be useful anymore.

Then the actual proposition uses update-ref for normal references, and a file deletion when the SHA1 is == ""

Tested with Git 2.17/2.30 and Gitea 1.15 and Dev version.
Do you agree with this ?